### PR TITLE
Fixes perf issue when Split results in 410 when trying to get latestN in Spark partitioner

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**
# Description
When using Spark Streaming to process a container's ChangeFeed there is a possible perf issue where data gets reprocessed (from beginning instead of previously processed LSN. This can happen when the Spark driver while calculating the Spark partitions for the next micro batch iteration receives a 410/* - in this case the StartLSN will be 0 instead of at least the LSN from the previous Offset. This change fixes this and makes sure to at least use the previously processed LSN.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PartitionMetadata.scala [hunk 1]
```diff
@@ -128,11 +128,26 @@ private[cosmos] case class PartitionMetadata
 
   def getAvgItemsPerLsn: Double = {
     if (this.firstLsn.isEmpty) {
-      math.max(1d, this.documentCount.toDouble / this.latestLsn)
-    } else if (this.documentCount == 0 || (this.latestLsn - this.firstLsn.get) <= 0) {
+      math.max(1d, this.documentCount.toDouble / this.getAndValidateLatestLsn())
+    } else if (this.documentCount == 0 || (this.getAndValidateLatestLsn() - this.firstLsn.get) <= 0) {
```
Comment (ealsur) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PartitionMetadata.scala:132]:
```suggestion
    } else if (this.documentCount == 0 || (this.getAndValidateLatestLsn - this.firstLsn.get) <= 0) {
```

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPartitionPlanner.scala [hunk 2]
```diff
@@ -543,29 +543,35 @@ private object CosmosPartitionPlanner extends BasicLoggingTrait {
...
+          case _: ReadAllAvailable => metadata.getAndValidateLatestLsn
           case maxRowsLimit: ReadMaxRows =>
             if (totalWeightedLsnGap.get <= maxRowsLimit.maxRows) {
+              val effectiveLatestLsn = metadata.getAndValidateLatestLsn
               if (isDebugLogEnabled) {
                 val calculateDebugLine = s"calculateEndLsn (feedRange: ${metadata.feedRange}) - avg. Docs " +
                   s"per LSN: ${metadata.getAvgItemsPerLsn} documentCount ${metadata.documentCount} firstLsn " +
-                  s"${metadata.firstLsn} latestLsn ${metadata.latestLsn} startLsn ${metadata.startLsn} weightedGap " +
-                  s"${metadata.getWeightedLsnGap} effectiveEndLsn ${metadata.latestLsn} maxRows ${maxRowsLimit.maxRows}"
+                  s"${metadata.firstLsn} latestLsn ${metadata.latestLsn} latestLsn ${metadata.latestLsn} " +
```
[the remark this answers is not in the document]
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosPartitionPlanner.scala:553]: Fixed in next iteration

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PartitionMetadata.scala [hunk 3]
```diff
@@ -128,11 +128,26 @@ private[cosmos] case class PartitionMetadata
 
   def getAvgItemsPerLsn: Double = {
     if (this.firstLsn.isEmpty) {
-      math.max(1d, this.documentCount.toDouble / this.latestLsn)
-    } else if (this.documentCount == 0 || (this.latestLsn - this.firstLsn.get) <= 0) {
+      math.max(1d, this.documentCount.toDouble / this.getAndValidateLatestLsn())
```
[the remark this answers is not in the document]
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PartitionMetadata.scala:131]: No - IntelliJ adds it by default - missed to fix it. Fixed in next iteration.

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PartitionMetadata.scala [hunk 4]
```diff
@@ -128,11 +128,26 @@ private[cosmos] case class PartitionMetadata
...
+    } else if (this.documentCount == 0 || (this.getAndValidateLatestLsn() - this.firstLsn.get) <= 0) {
       1d
     } else {
-      this.documentCount.toDouble / (this.latestLsn - this.firstLsn.get)
+      this.documentCount.toDouble / (this.getAndValidateLatestLsn() - this.firstLsn.get)
```
[the remark this answers is not in the document]
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PartitionMetadata.scala:135]: Fixed in next iteration

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PartitionMetadata.scala [hunk 5]
```diff
@@ -128,11 +128,26 @@ private[cosmos] case class PartitionMetadata
...
+    }
+  }
+
+  def getAndValidateLatestLsn(): Long = {
+    if (this.latestLsn == 0) {
+      // latestLsn == 0 but startLsn > 0 means there was an issue where change feed continuation
+      // was null - endLsn created here will be used as the startLsn for the next micro batch iteration
+      // so it should never be smaller than startLsn
+      math.max(this.latestLsn, this.startLsn)
```
[the remark this answers is not in the document]
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PartitionMetadata.scala:144]: Yeah - defense in depth - but in theory startLsn should be >= 0 - can remove this - probably just causes confusion.

[the remark this answers is not in the document]
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PartitionMetadata.scala:144]: Fixed in next iteration

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PartitionMetadata.scala (hunk 1 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/PartitionMetadata.scala:132]: Fixed in next iteration

**Pull request merged**